lighthouse-core/gather/gatherers/dobetterweb/domstats.js

```diff
@@ -124,14 +64,13 @@ function getDOMStats(element, deep = true) {
   return {
     depth: {
       max: result.maxDepth,
-      pathToElement: elementPathInDOM(deepestElement),
-      // ignore style since it will provide no additional context, and is often long
-      snippet: getOuterHTMLSnippet(deepestElement, ['style']),
+      devtoolsNodePath: getNodePath(deepestElement),
```

Comment:
I was hoping to use getNodeDetails to get all the info. But, interestingly while running a smoke test (dbw), there was a case when the nodeSelector function failed because the node/element had an undefined tagName? So, I decided to fall back to just gathering the devtoolsNodePath & path.

Reply:
It looks like that is because the element being resolved is a `ShadowRoot` not an `Element`. `ShadowRoot` does not have a `tagName` field so `getNodeDetails` runs into an error in the functions that use `tagName`.

I think `getNodeDetails` can work if you pass in the shadow root host when `parentWithMostChildren` is a `ShadowRoot` not an `Element`. You could also change `getNodeSelector` and `getNodeLabel` to handle a `ShadowRoot` being passed in. `getOuterHTMLSnippet` and `getNodePath` already do this which explains why those functions were working on their own:
https://github.com/GoogleChrome/lighthouse/blob/8095bec9d64e5d7f094374009eacb4a3a504a016/lighthouse-core/lib/page-functions.js#L115-L125

Reply:
Thanks for the explanation this worked! I'm wondering, though, why the `shadowRoot` case might have not been considered for the other functions (including `getBoundingClientRect(element)`), if it were on purpose or just a case that hadn't been encountered before - which I'm thinking might be the case.